firmware-management/mbl-watchdog-init/mbl-watchdog-init.c

```diff
@@ -0,0 +1,152 @@
+#include <string.h>
+#include <sys/ioctl.h>
+#include <unistd.h>
+
+//============================================================================
+// Error handling and logging convenience functions
+int is_err(const int ret)
+{
+    return ret == -1;
+}
+
+
+int log_error(const char *message)
+{
+    int err_code = errno;
+    char *prefix = "WATCHDOG-ERROR:";
+    char result[strlen(prefix) + strlen(message)];
+
+    snprintf(result, sizeof(result), "%s %s", prefix, message);
+    perror(result);
+
+    return err_code;
+}
+
+
+void log_warning(const char *message)
+{
+    char *prefix = "WATCHDOG-WARNING:";
+    printf("%s %s\n", prefix, message);
+}
+
+
+void log_info(const char *message)
+{
+    char *prefix = "WATCHDOG-INFO:";
+    printf("%s %s\n", prefix, message);
+}
+
+//============================================================================
+// Watchdog functions
+int set_watchdog_timeout(const int watchdog_fd, int *timeout)
+{
+    return ioctl(watchdog_fd, WDIOC_SETTIMEOUT, timeout);
+}
+
+
+int get_last_boot_status(const int watchdog_fd, int *flags)
+{
+    return ioctl(watchdog_fd, WDIOC_GETBOOTSTATUS, flags);
+}
+
+
+int print_last_boot_reason(const int boot_status)
+{
+    switch (boot_status)
+    {
+        case WDIOF_OVERHEAT:
+            log_info("The last reboot was caused by the CPU overheating.\n");
+            return 0;
+        case WDIOF_CARDRESET:
+            log_info("The last reboot was caused by a watchdog reset.\n");
+            return 0;
+    }
+    return -1;
+}
+
+//============================================================================
+// Main entry point
+int main(int argc, char **argv)
+{
+    int timeout = 0;
+    char* wdog_filename = "";
+
+    // Parse command line arguments
+    int current_opt;
+    int optindex;
+    struct option longopts[] = {
+        {"timeout", required_argument, NULL, 't'},
+        {"wdog-fname", required_argument, NULL, 'w'},
+    };
+
+    while ((current_opt = getopt_long(argc, argv, "tw:", longopts, &optindex)) != -1)
+    {
+        switch (current_opt)
+        {
+            case 't':
+                timeout = atoi(optarg);
+                break;
+            case 'w':
+                wdog_filename = optarg;
+                break;
+            default:
+                log_warning("Only --timeout and --wdog-fname should be provided. Exiting.");
+                return EXIT_FAILURE;
+        }
+    }
+
+    const int wdog_fd = open(wdog_filename, O_WRONLY);
+    if (is_err(wdog_fd))
+    {
+        return log_error("Failed to open watchdog device file");
+    }
+
+    int flags;
+    const int gs_ret = get_last_boot_status(wdog_fd, &flags);
+    if (is_err(gs_ret))
+    {
+        return log_error("Failed to get the last boot status");
+    }
+    else
+    {
+        const int lb_ret = print_last_boot_reason(flags);
+        if (is_err(lb_ret))
+        {
+            log_warning("Failed to determine last boot reason.");
+        }
+    }
```

Comment:
Can we encapsulate this logic inside the `get_last_boot_status`? It will make the code cleaner and easier to read.

Reply:
55d8c9d